File: __tests__/lib/session-retype.test.ts
// Tests du changement de type d'une séance enregistrée (lib/session-retype.ts).
// Cœur pur : décider si le changement est permis, et recalculer ce qui dépend du
// type. L'écriture en base (db.changeSessionType) n'est pas couverte ici.
import { canRetype, retypeChanges } from '@/lib/session-retype';
import type { ActivityType, Session } from '@/lib/types';

const PROFILE = { weightKg: 70, maxHr: 190 };

/** Sortie d'une heure, 20 km, 20 km/h de moyenne — de quoi comparer les barèmes. */
const session = (overrides: Partial<Session> = {}): Session => ({
  id: 1,
  type: 'velo',
  startedAt: 1_700_000_000_000,
  endedAt: 1_700_003_600_000,
  durationSec: 3600,
  movingTimeSec: 3600,
  notes: null,
  avgHr: null,
  maxHr: null,
  distanceM: 20_000,
  avgSpeedKmh: 20,
  maxSpeedKmh: 30,
  elevationGainM: 100,
  avgCadence: 85,
  maxCadence: 110,
  calories: 500,
  source: null,
  externalId: null,
  ...overrides,
});

describe('canRetype', () => {
  it('autorise les échanges entre activités tracées', () => {
    expect(canRetype('velo', 'marche')).toBe(true);
    expect(canRetype('velo', 'course')).toBe(true);
    expect(canRetype('marche', 'course')).toBe(true);
    expect(canRetype('course', 'velo')).toBe(true);
  });

  it('refuse la musculation dans les deux sens', () => {
    expect(canRetype('muscu', 'velo')).toBe(false);
    expect(canRetype('velo', 'muscu')).toBe(false);
    expect(canRetype('muscu', 'course')).toBe(false);
  });

  it('refuse un changement vers le type courant', () => {
    for (const t of ['velo', 'course', 'marche', 'muscu'] as ActivityType[]) {
      expect(canRetype(t, t)).toBe(false);
    }
  });
});

describe('retypeChanges', () => {
  it('renvoie null quand le changement n’est pas permis', () => {
    expect(retypeChanges(session(), 'velo', PROFILE)).toBeNull();
    expect(retypeChanges(session(), 'muscu', PROFILE)).toBeNull();
    expect(retypeChanges(session({ type: 'muscu' }), 'velo', PROFILE)).toBeNull();
  });

  it('porte le nouveau type', () => {
    expect(retypeChanges(session(), 'marche', PROFILE)!.type).toBe('marche');
  });

  it('ré-estime les calories avec le barème de la nouvelle activité', () => {
    const s = session();
    const versCourse = retypeChanges(s, 'course', PROFILE)!;
    const versMarche = retypeChanges(s, 'marche', PROFILE)!;
    // À 20 km/h : courir coûte beaucoup plus que pédaler, marcher (extrapolation
    // plate au-delà de la table) reste au-dessus du vélo mais loin de la course.
    expect(versCourse.calories).toBeGreaterThan(s.calories!);
    expect(versCourse.calories).toBeGreaterThan(versMarche.calories);
    // La valeur stockée d'origine n'entre jamais dans le nouveau calcul.
    const memeSeanceAutreCalories = session({ calories: 99_999 });
    expect(retypeChanges(memeSeanceAutreCalories, 'course', PROFILE)!.calories).toBeCloseTo(
      versCourse.calories,
      5,
    );
  });

  it('efface la cadence dès qu’on passe à pied', () => {
    for (const to of ['course', 'marche'] as ActivityType[]) {
      const changes = retypeChanges(session(), to, PROFILE)!;
      expect(changes.avgCadence).toBeNull();
      expect(changes.maxCadence).toBeNull();
    }
  });

  it('conserve la cadence en revenant au vélo', () => {
    const s = session({ type: 'course', avgCadence: 80, maxCadence: 95 });
    const changes = retypeChanges(s, 'velo', PROFILE)!;
    expect(changes.avgCadence).toBe(80);
    expect(changes.maxCadence).toBe(95);
  });

  it('compte le temps EN MOUVEMENT quand il est connu', () => {
    const avecArrets = session({ durationSec: 7200, movingTimeSec: 3600 });
    const sansArrets = session({ durationSec: 3600, movingTimeSec: null });
    expect(retypeChanges(avecArrets, 'course', PROFILE)!.calories).toBeCloseTo(
      retypeChanges(sansArrets, 'course', PROFILE)!.calories,
      5,
    );
  });

  it('retombe sur la durée totale sans temps en mouvement', () => {
    const court = session({ durationSec: 1800, movingTimeSec: null });
    const long = session({ durationSec: 3600, movingTimeSec: null });
    expect(retypeChanges(long, 'course', PROFILE)!.calories).toBeGreaterThan(
      retypeChanges(court, 'course', PROFILE)!.calories,
    );
  });

  it('utilise la FC max du PROFIL, pas celle de la séance', () => {
    // La FC max de la séance est une mesure ; celle du profil est la référence
    // d'intensité du modèle cardio. Les confondre fausserait l'estimation.
    const a = session({ avgHr: 150, maxHr: 175 });
    const b = session({ avgHr: 150, maxHr: 200 });
    expect(retypeChanges(a, 'course', PROFILE)!.calories).toBeCloseTo(
      retypeChanges(b, 'course', PROFILE)!.calories,
      5,
    );

    const profilPlusHaut = { weightKg: 70, maxHr: 210 };
    expect(retypeChanges(a, 'course', profilPlusHaut)!.calories).not.toBeCloseTo(
      retypeChanges(a, 'course', PROFILE)!.calories,
      5,
    );
  });

  it('tient compte du poids du profil', () => {
    const leger = retypeChanges(session(), 'course', { weightKg: 55, maxHr: 190 })!;
    const lourd = retypeChanges(session(), 'course', { weightKg: 95, maxHr: 190 })!;
    expect(lourd.calories).toBeGreaterThan(leger.calories);
  });
});

File: app/src/test/java/ovh/battistella/elan/domain/SessionRetypeTest.kt
// Tests du changement de type d'une séance enregistrée : décider si le
// changement est permis, et recalculer ce qui dépend du type.
package ovh.battistella.elan.domain

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class SessionRetypeTest {
    private val profile = Profile(weightKg = 70.0, maxHr = 190.0)

    /** Sortie d'une heure, 20 km, 20 km/h de moyenne — de quoi comparer les barèmes. */
    private fun session(
        type: ActivityType = ActivityType.VELO,
        durationSec: Int = 3600,
        movingTimeSec: Int? = 3600,
        avgHr: Double? = null,
        maxHr: Double? = null,
        avgCadence: Double? = 85.0,
        maxCadence: Double? = 110.0,
        calories: Double? = 500.0,
    ) = Session(
        id = 1,
        type = type,
        startedAt = 1_700_000_000_000,
        endedAt = 1_700_003_600_000,
        durationSec = durationSec,
        movingTimeSec = movingTimeSec,
        notes = null,
        avgHr = avgHr,
        maxHr = maxHr,
        distanceM = 20_000.0,
        avgSpeedKmh = 20.0,
        maxSpeedKmh = 30.0,
        elevationGainM = 100.0,
        avgCadence = avgCadence,
        maxCadence = maxCadence,
        calories = calories,
        source = null,
        externalId = null,
    )

    @Test
    fun `canRetype autorise les échanges entre activités tracées`() {
        assertTrue(canRetype(ActivityType.VELO, ActivityType.MARCHE))
        assertTrue(canRetype(ActivityType.VELO, ActivityType.COURSE))
        assertTrue(canRetype(ActivityType.MARCHE, ActivityType.COURSE))
        assertTrue(canRetype(ActivityType.COURSE, ActivityType.VELO))
    }

    @Test
    fun `canRetype refuse la musculation dans les deux sens`() {
        assertFalse(canRetype(ActivityType.MUSCU, ActivityType.VELO))
        assertFalse(canRetype(ActivityType.VELO, ActivityType.MUSCU))
        assertFalse(canRetype(ActivityType.MUSCU, ActivityType.COURSE))
    }

    @Test
    fun `canRetype refuse un changement vers le type courant`() {
        for (t in ActivityType.entries) assertFalse(canRetype(t, t))
    }

    @Test
    fun `renvoie null quand le changement n'est pas permis`() {
        assertNull(retypeChanges(session(), ActivityType.VELO, profile))
        assertNull(retypeChanges(session(), ActivityType.MUSCU, profile))
        assertNull(retypeChanges(session(type = ActivityType.MUSCU), ActivityType.VELO, profile))
    }

    @Test
    fun `porte le nouveau type`() {
        assertEquals(ActivityType.MARCHE, retypeChanges(session(), ActivityType.MARCHE, profile)!!.type)
    }

    @Test
    fun `ré-estime les calories avec le barème de la nouvelle activité`() {
        val s = session()
        val versCourse = retypeChanges(s, ActivityType.COURSE, profile)!!
        val versMarche = retypeChanges(s, ActivityType.MARCHE, profile)!!
        // À 20 km/h : courir coûte beaucoup plus que pédaler, marcher (extrapolation
        // plate au-delà de la table) reste au-dessus du vélo mais loin de la course.
        assertTrue(versCourse.calories > s.calories!!)
        assertTrue(versCourse.calories > versMarche.calories)
        // La valeur stockée d'origine n'entre jamais dans le nouveau calcul.
        val memeSeanceAutreCalories = session(calories = 99_999.0)
        assertEquals(versCourse.calories, retypeChanges(memeSeanceAutreCalories, ActivityType.COURSE, profile)!!.calories, 1e-5)
    }

    @Test
    fun `efface la cadence dès qu'on passe à pied`() {
        for (to in listOf(ActivityType.COURSE, ActivityType.MARCHE)) {
            val changes = retypeChanges(session(), to, profile)!!
            assertNull(changes.avgCadence)
            assertNull(changes.maxCadence)
        }
    }

    @Test
    fun `conserve la cadence en revenant au vélo`() {
        val s = session(type = ActivityType.COURSE, avgCadence = 80.0, maxCadence = 95.0)
        val changes = retypeChanges(s, ActivityType.VELO, profile)!!
        assertEquals(80.0, changes.avgCadence!!, 0.0)
        assertEquals(95.0, changes.maxCadence!!, 0.0)
    }

    @Test
    fun `compte le temps EN MOUVEMENT quand il est connu`() {
        val avecArrets = session(durationSec = 7200, movingTimeSec = 3600)
        val sansArrets = session(durationSec = 3600, movingTimeSec = null)
        assertEquals(
            retypeChanges(sansArrets, ActivityType.COURSE, profile)!!.calories,
            retypeChanges(avecArrets, ActivityType.COURSE, profile)!!.calories,
            1e-5,
        )
    }

    @Test
    fun `retombe sur la durée totale sans temps en mouvement`() {
        val court = session(durationSec = 1800, movingTimeSec = null)
        val long = session(durationSec = 3600, movingTimeSec = null)
        assertTrue(
            retypeChanges(long, ActivityType.COURSE, profile)!!.calories >
                retypeChanges(court, ActivityType.COURSE, profile)!!.calories,
        )
    }

    @Test
    fun `utilise la FC max du PROFIL, pas celle de la séance`() {
        val a = session(avgHr = 150.0, maxHr = 175.0)
        val b = session(avgHr = 150.0, maxHr = 200.0)
        assertEquals(
            retypeChanges(b, ActivityType.COURSE, profile)!!.calories,
            retypeChanges(a, ActivityType.COURSE, profile)!!.calories,
            1e-5,
        )

        val profilPlusHaut = Profile(weightKg = 70.0, maxHr = 210.0)
        assertNotEquals(
            retypeChanges(a, ActivityType.COURSE, profile)!!.calories,
            retypeChanges(a, ActivityType.COURSE, profilPlusHaut)!!.calories,
            1e-5,
        )
    }

    @Test
    fun `tient compte du poids du profil`() {
        val leger = retypeChanges(session(), ActivityType.COURSE, Profile(weightKg = 55.0, maxHr = 190.0))!!
        val lourd = retypeChanges(session(), ActivityType.COURSE, Profile(weightKg = 95.0, maxHr = 190.0))!!
        assertTrue(lourd.calories > leger.calories)
    }
}
